File: fudi-store/src/test/java/com/romellfudi/sharepreference/JsonUtilTest.java
package com.romellfudi.sharepreference;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.romellfudi.sharepreference.resources.ObjectBean;
import com.romellfudi.sharepreference.resources.ObjectDetailBean;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicReference;

import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.doAnswer;
import static org.powermock.api.mockito.PowerMockito.when;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.mockito.Mockito.times;
import static org.hamcrest.MatcherAssert.assertThat;

/**
 * @version 1.0
 * @autor Romell Domínguez
 * @date 9/30/18
 */
@RunWith(PowerMockRunner.class)
public class JsonUtilTest {

    @Mock
    Context context;

    @Mock
    SharedPreferences sharedPref;

    @Mock
    SharedPreferences.Editor editor;

    AtomicReference storeValue = new AtomicReference();

    @Captor
    ArgumentCaptor<String> keyCaptor, JSONvalueCaptor;

    @Before
    public void prepareMocks() {
        when(context.getPackageName()).thenReturn(getClass().getPackage().toString());
        when(context.getSharedPreferences(any(String.class), any(Integer.class)))
                .thenReturn(sharedPref);
        when(sharedPref.edit()).thenReturn(editor);
        SharePreferenced.init(context);
        doAnswer(new Answer() {
            @Override
            public Object answer(InvocationOnMock invocation) throws Throwable {
                storeValue.set(invocation.getArguments()[0]);
                if (storeValue.get().equals(keyCaptor.getValue()))
                    return JSONvalueCaptor.getValue();
                else return null;
            }
        }).when(sharedPref).getString(any(String.class), eq(""));
    }

    @Test
    public void testSimpleObjectToJson() throws Exception {
        ObjectDetailBean objectDetailBean = new ObjectDetailBean("234");
        objectDetailBean.setName("fudi");
        ArrayList<String> data = new ArrayList<>();
        data.add("hello");
        data.add("wait");
        data.add("bye");
        objectDetailBean.setDetails(data);
        SessionObj.setObjectDetailBeanCurrent(objectDetailBean);
        verify(editor).putString(keyCaptor.capture(), JSONvalueCaptor.capture());

        ObjectDetailBean storedValue = SessionObj.getObjectDetailBeanCurrent();
        assertThat(storedValue,is(notNullValue()));
        assertThat(storedValue,not(equalTo(objectDetailBean)));
        assertThat(storedValue.getName(),is(equalTo(objectDetailBean.getName())));
        assertThat(storedValue.getDetails(),is(equalTo(objectDetailBean.getDetails())));
        assertThat(storedValue.getId(),not(equalTo(objectDetailBean.getId())));

    }

    @Test
    public void testComplexObjectToJson() throws Exception {
        ArrayList<ObjectDetailBean> detailBeans = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            ArrayList<String> info = new ArrayList<String>();
            info.add("aa");
            info.add("bbbb");
            info.add("c");
            ObjectDetailBean detailBean = new ObjectDetailBean(info);
            detailBean.setName("id-"+i);

            SessionObj.setObjectDetailBeanCurrent(detailBean);
            verify(editor,times(i+1))
                    .putString(keyCaptor.capture(), JSONvalueCaptor.capture());
            ObjectDetailBean storedValue = SessionObj.getObjectDetailBeanCurrent();

            assertThat(storedValue,is(notNullValue()));
            assertThat(storedValue,not(equalTo(detailBean)));
            assertThat(storedValue.getName(),is(equalTo(detailBean.getName())));
            assertThat(storedValue.getDetails(),is(equalTo(detailBean.getDetails())));

            detailBeans.add(storedValue);
        }
        ObjectBean objectBeanMain = new ObjectBean("Main");
        objectBeanMain.setObjectDetailBeans(detailBeans);

        SessionObj.currentObject(objectBeanMain);
        verify(editor,times(detailBeans.size()+1))
                .putString(keyCaptor.capture(), JSONvalueCaptor.capture());
        ObjectBean storedObjectBeanMain = SessionObj.currentObject();

        assertThat(storedObjectBeanMain,is(notNullValue()));
        for (int i =0; i<storedObjectBeanMain.getObjectDetailBeans().size();i++) {
            ObjectDetailBean storedValue = storedObjectBeanMain.getObjectDetailBeans().get(i);
            ObjectDetailBean detailBean = detailBeans.get(i);
            assertThat(storedValue,is(notNullValue()));
            assertThat(storedValue,not(equalTo(detailBean)));
            assertThat(storedValue.getName(),is(equalTo(detailBean.getName())));
            assertThat(storedValue.getDetails(),is(equalTo(detailBean.getDetails())));
        }
        assertThat(storedObjectBeanMain.getData(),is(equalTo(objectBeanMain.getData())));
    }

}

File: fudi-store-kotlin/src/test/java/com/romellfudi/sharepreference/JsonUtilTest.kt
/*
 * Copyright (c) 2021. BoostTag E.I.R.L. Romell D.Z.
 * All rights reserved
 * portfolio.romellfudi.com
 */

package com.romellfudi.sharepreference

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.romellfudi.sharepreference.resources.NonAnnotation
import com.romellfudi.sharepreference.resources.NonDataClass
import com.romellfudi.sharepreference.resources.ObjectBean
import com.romellfudi.sharepreference.resources.ObjectDetailBean
import io.mockk.MockKAnnotations
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.slot
import io.mockk.verify
import org.hamcrest.CoreMatchers.*
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.FixMethodOrder
import org.junit.Test
import org.junit.runners.MethodSorters
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.powermock.core.classloader.annotations.PrepareForTest
import java.util.concurrent.atomic.AtomicReference

/**
 * Created by romelldominguez on 5/15/16.
 */
@PrepareForTest(Uri::class)
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
class JsonUtilTest {

    @Mock
    internal val context: Context = mock()

    @Mock
    internal val sharedPref: SharedPreferences = mock()

    @RelaxedMockK
    lateinit var editor: SharedPreferences.Editor

    private var storeValue = AtomicReference<String>()

    private val keyCaptor = slot<String>()
    private val JSONvalueCaptor = slot<String>()

    @Before
    fun prepareMocks() {
        MockKAnnotations.init(this, relaxUnitFun = true)
        `when`(context.packageName).thenReturn(javaClass.getPackage()!!.toString())
        `when`(context.getSharedPreferences(any(), any())).thenReturn(sharedPref)
        `when`(sharedPref.edit()).thenReturn(editor)
        SharePreferenced.init(context)
        whenever(sharedPref.getString(any(), eq("")))
                .thenAnswer {
                    storeValue.set(it.arguments[0] as String?)
                    if (storeValue.get() == keyCaptor.captured) JSONvalueCaptor.captured else null
                }
    }

    @Test
    @Throws(Exception::class)
    fun nonAnnotationTest() {
        val nonAnnotationBean = NonAnnotation("234")
                .apply {
                    name = "fudi"
                    details = arrayListOf("hello", "wait", "bye")
                }
        assertThat(nonAnnotationBean::class.isData, `is`(true))
        assertThat(nonAnnotationBean::class.annotations.find { it is Fudi } as? Fudi,
                `is`(nullValue()))
        nonAnnotationBean.save()
        val storedValue = NonAnnotation()
        storedValue.load()
        assertThat(storedValue == nonAnnotationBean, `is`(false))
    }

    @Test
    @Throws(Exception::class)
    fun nonDataClassTest() {
        val nonDataClass = NonDataClass("234")
                .apply {
                    name = "fudi"
                    details = arrayListOf("hello", "wait", "bye")
                }
        assertThat(nonDataClass::class.isData, `is`(false))
        assertThat(nonDataClass::class.annotations.find { it is Fudi } as? Fudi,
                `is`(notNullValue()))
        nonDataClass.save()
        val storedValue = NonDataClass()
        storedValue.load()
        assertThat(storedValue == nonDataClass, `is`(false))
    }

    @Test
    @Throws(Exception::class)
    fun testSimpleObjectToJson() {
        val objectDetailBean = ObjectDetailBean("234")
                .apply {
                    name = "fudi"
                    details = arrayListOf("hello", "wait", "bye")
                }
        assertThat(objectDetailBean::class.isData, `is`(true))
        assertThat(objectDetailBean::class.annotations.find { it is Fudi } as? Fudi,
                `is`(notNullValue()))
        objectDetailBean.save()
        verify { editor.putString(capture(keyCaptor), capture(JSONvalueCaptor)) }
        val storedValue = ObjectDetailBean()
        storedValue.load()
        assertThat<ObjectDetailBean>(objectDetailBean, `is`(notNullValue()))
        assertThat<ObjectDetailBean>(storedValue, `is`(notNullValue()))
        assertThat(storedValue, equalTo(objectDetailBean))
    }

    @Test
    @Throws(Exception::class)
    fun testComplexObjectToJson() {
        val detailBeans = ArrayList<ObjectDetailBean>()
        (0..4).forEach {
            val detailBean = ObjectDetailBean().apply {
                name = "id-$it"
                details = arrayListOf("aa", "bbbb", "c")
            }
            assertThat(detailBean::class.isData, `is`(true))
            detailBean.save()
            verify(exactly = it + 1) {
                editor.putString(capture(keyCaptor), capture(JSONvalueCaptor))
            }
            val storedValue = ObjectDetailBean()
            storedValue.load()
            assertThat<ObjectDetailBean>(storedValue, `is`(notNullValue()))
            assertThat(storedValue, equalTo(detailBean))
            detailBeans.add(storedValue)
        }
        val objectBeanMain = ObjectBean("Main")
        assertThat(objectBeanMain::class.isData, `is`(true))
        assertThat(objectBeanMain::class.annotations.find { it is Fudi } as? Fudi,
                `is`(notNullValue()))

        objectBeanMain.save()

        verify(exactly = detailBeans.size + 1) {
            editor.putString(capture(keyCaptor), capture(JSONvalueCaptor))
        }
        val storedObjectBeanMain = ObjectBean("Main")
        storedObjectBeanMain.load()

        assertThat(storedObjectBeanMain, `is`(notNullValue()))
        storedObjectBeanMain.objectDetailBeans?.forEachIndexed { index, objectDetailBean ->
            assertThat(objectDetailBean, `is`(notNullValue()))
            assertThat(detailBeans[index], `is`(notNullValue()))
            assertThat(objectDetailBean, equalTo(detailBeans[index]))
        }
        assertThat(storedObjectBeanMain.data, `is`(equalTo(objectBeanMain.data)))
    }

}